sound/soc/sof/sof-dai-clk.c

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Brent Lu <brent.lu@intel.com>
+//
+// DAI clock layer to support explicit clock control in machine driver.
+//
+
+#include <linux/clk-provider.h>
+#include <linux/clkdev.h>
+#include <sound/sof/dai.h>
+#include "sof-priv.h"
+#include "sof-dai-clk.h"
+
+struct sof_clk {
+	struct clk_hw hw;
+	struct snd_sof_dev *sdev;
+	struct sof_dai_clk *dai_clk;
+};
+
+static int sof_clk_ctrl(struct snd_sof_dev *sdev, struct sof_dai_clk *dai_clk, int enable)
+{
+	struct sof_ipc_dai_clkctrl clkctrl;
+	struct sof_ipc_reply reply;
+	int ret;
+
+	memset(&clkctrl, 0, sizeof(clkctrl));
+
+	/* set IPC DAI clock control */
+	clkctrl.hdr.size = sizeof(clkctrl);
+	clkctrl.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CLKCTRL;
+
+	clkctrl.type = dai_clk->type;
+	clkctrl.dai_index = dai_clk->dai_index;
+
+	clkctrl.clk_en = enable ? 1 : 0;
+	clkctrl.clk_id = dai_clk->clk_id;
+
+	dev_dbg(sdev->dev, "clkctrl: SSP%d clk_en %d clk_id %d\n",
```

Comment:
is this clock SSP-only? If it is, it should be clear at the file / function level, if it isn't, it shouldn't be hard-coded here

Reply:
It might be better to be human readable:
`dev_dbg(sdev->dev, "clkctrl: %s (clkid: %d) - %s\n", dai_clk->name, dai_clk->clk_id, enable ? "Enable" : "Disable");`